packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts

```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
         TopicArn: this.props.topic.topicArn,
         Message: this.props.message.value,
         MessageStructure: this.props.messagePerSubscriptionType ? 'json' : undefined,
+        MessageAttributes: renderMessageAttributes(this.props.messageAttributes),
         Subject: this.props.subject,
       }),
     };
   }
 }
+
+interface MessageAttributeValue {
+  DataType: string;
+  StringValue?: string | null;
+  BinaryValue?: string;
+}
+
+function renderMessageAttributes(attributes?: { [key: string]: MessageAttribute }): any {
+  if (attributes === undefined) { return undefined; }
+  const renderedAttributes: { [key: string]: MessageAttributeValue } = {};
+  Object.entries(attributes).map(([key, val]) => {
+    renderedAttributes[key] = renderMessageAttributeValue(val);
+  });
+  return sfn.TaskInput.fromObject(renderedAttributes).value;
+}
+
+function renderMessageAttributeValue(attribute: MessageAttribute): MessageAttributeValue {
+  const dataType = attribute.dataType;
+  if (attribute.value instanceof sfn.TaskInput) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING, StringValue: attribute.value.value };
+  }
+
+  if (dataType === MessageAttributeDataType.BINARY) {
+    return { DataType: dataType, BinaryValue: `${attribute.value}` };
+  }
+
+  if (Token.isUnresolved(attribute.value)) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING, StringValue: attribute.value };
+  }
+
+  validateMessageAttribute(attribute);
+  if (Array.isArray(attribute.value)) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING_ARRAY, StringValue: JSON.stringify(attribute.value) };
+  }
+  const value = attribute.value;
+  if (typeof value === 'number') {
+    return { DataType: MessageAttributeDataType.NUMBER, StringValue: `${value}` };
+  } else {
+    return { DataType: MessageAttributeDataType.STRING, StringValue: `${value}` };
+  }
+}
+
+function validateMessageAttribute(attribute: MessageAttribute): void {
+  const dataType = attribute.dataType;
+  const value = attribute.value;
+  if (dataType === undefined) {
+    return;
+  }
+  if (Array.isArray(value)) {
+    if (dataType !== MessageAttributeDataType.STRING_ARRAY) {
+      throw new Error(`Unsupported SNS message attribute type: ${JSON.stringify(value)}`);
```

Comment:
For this and the other error messages below, can we use the following pattern:
> Requested SNS message attribute type was ${dataType} but value was of type [array/string/...]

Reply:
Is it worth trying to print the value? Right now the message will tell you the value doesn't match the type but no which value?

Reply:
Yes – will add this!